platform/engineering/design-docs/2021-05-18-forms-library.md

```diff
@@ -0,0 +1,852 @@
+  </>
+)
+```
+
+Notice how in the first example, the function must be called with the value
+after all optional error message override functions, but in the second example,
+Formik calls the function with the value for the field itself.
+
+**Use the field value in the message:**
+```js
+const validateField = fromSchema('path')
+  .maxLength(value => `You entered ${value.length} characters, which is too many.`);
+```
+
+##### Caveat
+This field-level validation doesn't ensure that the form data submitted to the
+API matches the same shape as the JSON schema defines. Because a mismatch in the
+shape of the form data and corresponding schema has nothing to do with what the
+user entered, this is a developer mistake and should be caught by developers and
+not user-facing validation errors.
+
+To clarify, take the following example:
+**Schema**
+```json
+{
+  "type": "object",
+  "properties": {
+    "email": { "type": "string" },
+    "email-confirmation": { "type": "string" },
+  }
+}
+```
+
+**Data**
+```js
+{
+  email: {
+    original: "foo@bar.com",
+    confirmation: "foo@bar.com"
+  }
+}
+```
+Formulate will not deliberately catch mistakes like this, but it should be
+reasonably difficult to make these mistakes when using the `fromSchema`
+validation since the field's `name` property should likely match the schema path
+used in validation.
+
+Implementing such a mistake would look something like this using the
+`fromSchema` validation:
+```jsx
+<StringField
+  name="email-confirmation"
+  validation={fromSchema('email.confirmation')} />
+```
+
+**Discussion:** Is not ensuring the shape of the data matches the schema a
+reasonable boundary for the library to stop at?
+
+**Discussion:** What tools might Formulate give to developers to help catch
+these kinds of data shape mismatch mistakes?
+
+### Code Location
+Formulate will live in its own repository.
+
+### Documentation
+Formulate documentation will follow the [diataxis framework](https://diataxis.fr/).
+
+The following documentation is considered essential for the successful adoption
+of Formulate.
+- [Tutorials](https://diataxis.fr/tutorials/)
+  - Getting started
+  - Building a multi-page form
+- [How-to guides](https://diataxis.fr/how-to-guides/)
+  - How to build a review page
+  - How to make array pages
+  - How to make single-page list loops
+  - How to build a custom field
+  - How to write a data migration
+- [Reference documentation](https://diataxis.fr/reference/) tooling
+  - See[Reference documentation](#reference-documentation) below
+- [Explanations](https://diataxis.fr/explanation/)
+  - Why keep form state in a React context for multi-page forms?
+
+All documentation will be located within the Formulate repo in the `docs/`
+directory. It is essential that documentation and code live side-by-side so the
+documentation may be updated at the same time as the code. Additionally, keeping
+all the documentation in `docs/` allows us to serve it simply using GitHub Pages.
```

Comment:
Will we be serving it on design.va.gov?

Reply:
No, probably not. I think design.va.gov is probably not the right place for all this developer-facing documentation. For simplicity, we'll stick to GitHub pages and link to that wherever it makes sense (like the project README).